File: app/src/main/java/com/rtukpe/fixtures/ui/competition/table/TableAdapter.java
package com.rtukpe.fixtures.ui.competition.table;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.rtukpe.fixtures.R;
import com.rtukpe.fixtures.data.model.LeagueTable;
import com.rtukpe.fixtures.utils.others.ImageUtils;
import com.rtukpe.fixtures.utils.others.RecyclerViewClickListener;

import java.util.ArrayList;
import java.util.Collection;

import butterknife.BindView;
import butterknife.ButterKnife;

public class TableAdapter extends RecyclerView.Adapter<TableAdapter.ViewHolder> {

    private ArrayList<LeagueTable.Standing> standings;
    private Context mContext;
    private RecyclerViewClickListener mRecyclerViewClickListener;

    public TableAdapter(@NonNull Context context) {
        this.mContext = context;
        this.standings = new ArrayList<>();
    }

    public ArrayList<LeagueTable.Standing> getStandings() {
        return standings;
    }

    public void setRecyclerViewClickListener(RecyclerViewClickListener mRecyclerViewClickListener) {
        this.mRecyclerViewClickListener = mRecyclerViewClickListener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(mContext).inflate(R.layout.layout_table_standing_item, parent, false);
        return new ViewHolder(view, mRecyclerViewClickListener);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.standingName.setText(standings.get(holder.getAdapterPosition()).getTeam());
        holder.standingIndex.setText(String.valueOf(standings.get(holder.getAdapterPosition()).getRank()));
        holder.standingPoints.setText(String.valueOf(standings.get(holder.getAdapterPosition()).getPoints()));
        holder.standingPlayed.setText(String.valueOf(standings.get(holder.getAdapterPosition()).getPlayedGames()));
        holder.standingGoalDifference.setText(String.valueOf(standings.get(holder.getAdapterPosition()).getGoalDifference()));
        if (standings.get(holder.getAdapterPosition()).getLogo() != null && standings.get(holder.getAdapterPosition()).getLogo().contains("svg")) {
            ImageUtils.loadSVG(mContext, standings.get(holder.getAdapterPosition()).getLogo(), holder.standingLogo);
        } else {
            ImageUtils.displayImageFromUrl(mContext, standings.get(holder.getAdapterPosition()).getLogo(), holder.standingLogo);
        }
    }

    @Override
    public int getItemCount() {
        return standings.size();
    }

    public LeagueTable.Standing getStandingAtPosition(int position) {
        return standings.get(position);
    }

    public LeagueTable.Standing getFirstStanding() {
        return standings.get(0);
    }

    public LeagueTable.Standing getLastStanding() {
        return standings.get(standings.size() - 1);
    }

    public void addStanding(LeagueTable.Standing fixture) {
        standings.add(fixture);
        notifyDataSetChanged();
    }

    public void addStandings(Collection<LeagueTable.Standing> fixtures) {
        this.standings.addAll(fixtures);
        notifyDataSetChanged();
    }

    public void clear() {
        standings.clear();
        notifyDataSetChanged();
    }

    protected class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        @BindView(R.id.standing_index)
        TextView standingIndex;

        @BindView(R.id.standing_logo)
        ImageView standingLogo;

        @BindView(R.id.standing_name)
        TextView standingName;

        @BindView(R.id.standing_played)
        TextView standingPlayed;

        @BindView(R.id.standing_goal_diff)
        TextView standingGoalDifference;

        @BindView(R.id.standing_points)
        TextView standingPoints;

        RecyclerViewClickListener recyclerViewClickListener;

        ViewHolder(View itemView, RecyclerViewClickListener recyclerViewClickListener) {
            super(itemView);

            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(this);
            this.recyclerViewClickListener = recyclerViewClickListener;
        }

        @Override
        public void onClick(View v) {
            recyclerViewClickListener.recyclerViewListClicked(v, getAdapterPosition());
        }
    }
}

File: app/src/main/java/com/rtukpe/fixtures/ui/competition/table/TableAdapter.kt
package com.rtukpe.fixtures.ui.competition.table

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import butterknife.BindView
import butterknife.ButterKnife
import com.rtukpe.fixtures.R
import com.rtukpe.fixtures.data.model.LeagueTable
import com.rtukpe.fixtures.utils.others.ImageUtils
import com.rtukpe.fixtures.utils.others.RecyclerViewClickListener
import java.util.*

class TableAdapter(private val mContext: Context) : RecyclerView.Adapter<TableAdapter.ViewHolder>() {

    val standings: ArrayList<LeagueTable.Standing> = ArrayList()
    lateinit var mRecyclerViewClickListener: RecyclerViewClickListener

    val firstStanding: LeagueTable.Standing
        get() = standings[0]

    val lastStanding: LeagueTable.Standing
        get() = standings[standings.size - 1]

    fun setRecyclerViewClickListener(mRecyclerViewClickListener: RecyclerViewClickListener) {
        this.mRecyclerViewClickListener = mRecyclerViewClickListener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(mContext).inflate(R.layout.layout_table_standing_item, parent, false)
        return ViewHolder(view, mRecyclerViewClickListener)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {

        with(standings[holder.adapterPosition]) {
            holder.standingName.text = team
            holder.standingIndex.text = rank.toString()
            holder.standingPoints.text = points.toString()
            holder.standingPlayed.text = playedGames.toString()
            holder.standingGoalDifference.text = goalDifference.toString()

            if (logo != null && logo!!.contains("svg")) {
                logo?.let { ImageUtils.loadSVG(mContext, it, holder.standingLogo) }
            } else {
                logo?.let { ImageUtils.displayImageFromUrl(mContext, it, holder.standingLogo) }
            }
        }
    }

    override fun getItemCount(): Int {
        return standings.size
    }

    fun getStandingAtPosition(position: Int): LeagueTable.Standing {
        return standings[position]
    }

    fun addStanding(fixture: LeagueTable.Standing) {
        standings.add(fixture)
        notifyDataSetChanged()
    }

    fun addStandings(fixtures: Collection<LeagueTable.Standing>) {
        this.standings.addAll(fixtures)
        notifyDataSetChanged()
    }

    fun clear() {
        standings.clear()
        notifyDataSetChanged()
    }

    inner class ViewHolder internal constructor(itemView: View, internal var recyclerViewClickListener: RecyclerViewClickListener)
        : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        @BindView(R.id.standing_index)
        lateinit var standingIndex: TextView

        @BindView(R.id.standing_logo)
        lateinit var standingLogo: ImageView

        @BindView(R.id.standing_name)
        lateinit var standingName: TextView

        @BindView(R.id.standing_played)
        lateinit var standingPlayed: TextView

        @BindView(R.id.standing_goal_diff)
        lateinit var standingGoalDifference: TextView

        @BindView(R.id.standing_points)
        lateinit var standingPoints: TextView

        init {
            ButterKnife.bind(this, itemView)
            itemView.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            recyclerViewClickListener.recyclerViewListClicked(v, adapterPosition)
        }
    }
}
